### PR TITLE
Add linkat support

### PR DESCRIPTION
**Pull request by LucaSeri**

### CODE REVIEW

**Comment by LucaSeri** [comment]
This is currently based on https://github.com/ocaml-multicore/ocaml-uring/pull/104

File: lib/uring/uring.mli [hunk 1]
```diff
@@ -145,6 +145,18 @@ val openat2 : 'a t ->
     @param perm sets the access control bits for newly created files (subject to the process's umask)
     @param resolve controls how the pathname is resolved. *)
 
+val linkat : 'a t -> ?target_dir_fd:Unix.file_descr -> ?link_dir_fd:Unix.file_descr -> follow:bool -> target:string -> link_name:string -> 'a -> 'a job option
+(** [link t ?follow ~target ~link_name] creates a new hard link. If the path of the link
```
Comment (talex5) [inline lib/uring/uring.mli:149]:
```suggestion
(** [linkat t ~follow ~target ~link_name] creates a new hard link. If the path of the link
```
I'm not keen on the `target` and `link` names. They would make sense for symlinks, but either could be described as the "target" or as the "link" here. The Linux man-page uses `old` and `new` which seems clearer to me.

File: lib/uring/uring.mli [hunk 2]
```diff
@@ -145,6 +145,18 @@ val openat2 : 'a t ->
...
+    already exists, then it is not overwritten.
+    @param target_dir_fd If provided and [target] is a relative path, it is interpreted
+                         relative to [target_dir].
+    @param link_dir_fd If provided and [link_name] is a relative path, it is interpreted
+                       relative to [link_dir].
```
Comment (talex5) [inline lib/uring/uring.mli:154]:
```suggestion
    @param target_dir_fd If provided and [target] is a relative path, it is interpreted
                         relative to [target_dir_fd].
    @param link_dir_fd If provided and [link_name] is a relative path, it is interpreted
                       relative to [link_dir_fd].
```

File: lib/uring/uring_stubs.c [hunk 3]
```diff
@@ -998,3 +998,45 @@ ocaml_uring_opcode_supported(value v_probe, value v_op)
...
+
+  int ret = io_uring_register_eventfd(ring, fd);
+  if (ret)
+    unix_error(-ret, "io_uring_register_eventfd", Nothing);
+
+  return Val_unit;
+}
+
+value /* noalloc */
+ocaml_uring_submit_linkat_native(value v_uring, value v_id, value target_dir, value v_target, value link_dir, value v_link_name, value follow) {
```
Comment (talex5) [inline lib/uring/uring_stubs.c:1015]:
```suggestion
ocaml_uring_submit_linkat_native(value v_uring, value v_id, value v_target_dir, value v_target, value v_link_dir, value v_link_name, value v_follow) {
```

**Review by talex5 — COMMENTED** [review]
Looks good - some minor comments in-line.

**Comment by talex5** [comment]
Would be good to add a test to `tests/main.md` too.

**Comment by talex5** [comment]
I pushed the suggested changes to your branch, and also added the other flag, plus some tests.

However, the tests revealed that `linkat` doesn't work with `follow` (or any other flags), due to https://github.com/axboe/liburing/issues/955.

**Comment by LucaSeri** [comment]
Sorry I couldn't work on this for a while. Thank you @talex5 for the additions and digging into this!

**Review by talex5 — APPROVED** [review]

File: lib/uring/uring.mli [hunk 4]
```diff
@@ -251,6 +251,32 @@ val openat2 : 'a t ->
...
+  flags:Linkat_flags.t ->
+  old_path:string ->
+  new_path:string ->
+  'a -> 'a job option
+(** [linkat t ~flags ~old_path ~new_path] creates a new hard link.
+
+    If [new_path] already exists then it is not overwritten.
+
+    @param old_dir_fd If provided and [old_path] is a relative path, it is interpreted relative to [old_dir_fd].
+    @param new_dir_fd If provided and [new_name] is a relative path, it is interpreted relative to [new_dir_fd].
```
Comment (avsm) [inline lib/uring/uring.mli:276]:
```suggestion
    @param new_dir_fd If provided and [new_path] is a relative path, it is interpreted relative to [new_dir_fd].
```

File: lib/uring/uring.mli [hunk 5]
```diff
@@ -251,6 +251,32 @@ val openat2 : 'a t ->
...
+    @param old_path Path to which we create the link.
```
Comment (avsm) [inline lib/uring/uring.mli:277]:
```suggestion
    @param old_path Destination path of the newly created link.
```

Comment (talex5) [reply lib/uring/uring.mli]: I don't like either wording - the newly created link doesn't have a "destination path". How about:
```
    @param old_path Path of the already-existing link.
    @param new_path Path for the newly created link.
```

Comment (avsm) [reply lib/uring/uring.mli]: Yes, that's better.

**Pull request merged**